### PR TITLE
Background job routine status

### PR DESCRIPTION
**Pull request by jpslav**
This PR adds infrastructure to support reporting of routine status when run as a background job.  Instead of relying on one of the existing `<background job system>-status` gems out there, this PR implements a simple status mechanism that is `ActiveJob` friendly and ties into the existing Lev infrastructure. 

Tests not included :-)

### CODE REVIEW

File: lib/lev/errors.rb [hunk 1]
```diff
@@ -4,12 +4,29 @@ module Lev
...
+
     def add(fail, args={})
       args[:kind] ||= :lev
       error = Error.new(args)
+
       return if ignored_error_procs.any?{|proc| proc.call(error)}
       self.push(error)
-      throw :fatal_errors_encountered if fail
+
+      routine_status.add_error(error, is_fatal: fail)
```
Comment (jpslav) [inline lib/lev/errors.rb]: Can't have named arguments since some projects use lev with ruby 1.9

Comment (jpslav) [inline lib/lev/errors.rb]: if you want you can replace `add_error` with `add_fatal_error` and `add_nonfatal_error` instead.

File: lib/lev/status.rb [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    end
+
+    STATUSES.each do |status|
+      define_method("#{status}?") do
+        get('status') == status
+      end
+    end
+
+    def set_status_progress(at, out_of)
+      if out_of.nil? && (at < 0 || at > 1)
```
Comment (jpslav) [inline lib/lev/status.rb]: This condition looks like it belongs in `prevent_faulty_arguments`?  Can maybe the `prevent_faulty_arguments` checks be moved to `set_status_progress`?  Or if not, change the name of `prevent_faulty_arguments` to be progress-specific?

File: lib/lev/errors.rb (hunk 1 above)
Comment (jpslav) [inline lib/lev/errors.rb]: duh, sorry, you're passing as options.  all good :-)

File: lib/lev/status.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    def set(incoming_hash)
+      if existing_settings = self.class.find(uuid)
+        incoming_hash = existing_settings.merge(incoming_hash)
+      end
+
+      self.class.store.write(status_key, incoming_hash.to_json)
+    end
+
+    def status_key
+      "#{Lev.configuration.status_store_namespace}:#{uuid}"
```
Comment (jpslav) [inline lib/lev/status.rb]: `self.status_key(uuid)`

File: README.md [hunk 4]
```diff
@@ -441,6 +441,49 @@ Lev.configure do |config|
...
+* `'status'` which returns one of the status strings shown above
+* `'errors'` which (if present) is an error of error hashes
+* `'uuid'` the UUID of the routine / status
+
+Other routine-specific keys (set with a `save` call) are also present.
+
+**Notes:**
+
+1. Don't try to write a status store that uses the ActiveRecord database, as the database changes would only be seen when the routine completes and its transaction is committed.
+2. Job killing hasn't been implemented yet, but shouldn't be too bad.  For routines run in a transaction (which are frankly the only ones you'd want to kill), we can likely kill them by raising `ActiveRecord::Rollback`.
```
Comment (Dantemss) [inline README.md]: Kevin pointed out to me that ActiveRecord::Rollback does absolutely nothing in nested transactions, so you might wanna use a different Exception to be safe.

**Comment by Dantemss** [comment]
Looking good.

File: README.md (hunk 4 above)
Comment (jpslav) [inline README.md]: Good point. Noted in ba7634ba4373b178131dc59288df2e51138c869b

**Pull request merged**